File: spark_basic/task2.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s._
import org.json4s.jackson.JsonMethods._
/*
object task2 {
  def main(arg: Array[String]): Unit={
    val input_partition = 5
    val spark = new SparkConf().setAppName("task2").setMaster("local[*]")
    val sc = new SparkContext(spark)
    val input_file = "/Users/wentaozhou/desktop/test_review.json"
    val rdd = sc.textFile(path = input_file)
    val textRDD = rdd.map{ row => val json_row = jsonStrToMap(row)
      (json_row)
    }
    val t1 = System.nanoTime
    val toptenbusiness = textRDD.map(row => List(row("business_id"),1)).reduceByKey((a,b) => a+b)
    val so = toptenbusiness.sortBy(row => (-row._2.toInt, row._1.toString())).take(10)
    val duration = (System.nanoTime - t1) / 1e9d
    val numofpartition = toptenbusiness.getNumPartitions
    val itemperpartition = toptenbusiness.glom().map(x => x.size).collect()
    println(numofpartition)
    itemperpartition.foreach(println)
    print(duration)
    val t2 = System.nanoTime
    val topten_custom = textRDD.map(row => (row("business_id"),1)).reduceByKey((a,b) => a+b, input_partition)
    val so2 = toptenbusiness.sortBy(row => (-row._2.toInt, row._1.toString())).take(10)
    val duration2 = (System.nanoTime - t2) / 1e9d
    print(duration2)

  }



  def jsonStrToMap(jsonStr: String): Map[String, Any] = {
    implicit val formats = org.json4s.DefaultFormats
    parse(jsonStr).extract[Map[String, Any]]
  }
}
*/

File: spark_basic/task2.py
# Wentao Zhou
# DSCI 553 - hw1, task2, spring 2021
# spark warm-up practice

from pyspark import SparkContext
import json
import os
import time
import sys

sc = SparkContext('local[*]', 'task2')
input_file_path = sys.argv[1]
output = sys.argv[2]
input_partition = int(sys.argv[3])
textRDD = sc.textFile(input_file_path).map(json.loads)
textRDD =  textRDD.map(lambda row: row["business_id"]).cache()
# ======================= exe time of default partition function ===========================
startTime = time.time()
toptenbusiness = textRDD\
        .map(lambda row: (row, 1)).reduceByKey(lambda a, b: a+b)
topten = toptenbusiness.takeOrdered(10, key=lambda x: (-x[1], x[0]))
endTime = time.time()
exe_time = endTime - startTime

numofpartition = toptenbusiness.getNumPartitions()
numitem = toptenbusiness.glom().map(len).collect()
# ======================= exe time of default partition function ===========================

# ======================= exe time of custom partition function ===========================
startTime = time.time()
toptenbusiness = textRDD\
        .map(lambda row: [row, 1]).reduceByKey(lambda a, b: a+b, input_partition, lambda id: hash(id))
topten = toptenbusiness.takeOrdered(10, key=lambda x: (-x[1], x[0]))
endTime = time.time()
my_exe_time = endTime - startTime
my_numofpartition = toptenbusiness.getNumPartitions()
my_numitem = toptenbusiness.glom().map(len).collect()
# ======================= exe time of custom partition function ===========================

res = {}
res["default"] = {
    "n_partition": numofpartition,
    "n_items": numitem,
    "exe_time": exe_time
}
res["customized"] = {
    "n_partition": my_numofpartition,
    "n_items": my_numitem,
    "exe_time": my_exe_time
}
with open(output, 'w') as output:
    json.dump(res, output)
output.close()